lib/Authentication/AuthTokens/ShibAuthToken.php

```diff
@@ -96,73 +96,61 @@ private function getAttributesInitToken(){
-                 $HTML .= "</div>";
-                 echo ($HTML);
-                 die();
-            }
-            $this->principal = $_SERVER['voPersonID'];
-            $this->userDetails = array('AuthenticationRealm' => array('EGI Proxy IdP'));
-            return;
-            */
 
-            if( empty($_SERVER['voPersonID'])){// || empty($_SERVER['displayName']) ){
-                die('Did not recieve required attributes from the EGI Proxy Identity Provider to complete authentication, please contact gocdb-admins');
+        $configService = \Factory::getConfigService();
+        $identityProviders = $configService->getIdentityProvidersInfo();
+
+        foreach ($identityProviders as $provider) {
+            if ($provider['idp'] === $idp) {
+                $name = $provider['name'];
+                $helpUrl = $provider['help_url'] ?? '#';
+
+                if (empty($_SERVER['voPersonID'])) {
+                    die(
+                        "Did not receive required attributes from the IDP $name to "
+                        . "complete authentication. Please contact gocdb-admins."
+                    );
+                }
+
+                if (empty($_SERVER['entitlement'])) {
+                    die(
+                        "Did not receive the required entitlement attribute from "
+                        . "the IDP $name. Please contact gocdb-admins."
+                    );
+                }
+
+                if (!empty($provider['required_groups'])) {
+                    $entitlementValues = explode(
+                        ';', $_SERVER['entitlement']
+                    );
+
+                    if (!array_intersect(
+                        $entitlementValues, $provider['required_groups']
+                    )) {
+                        $HTML = "<ul>"
+                            . "<li>Login requires the entitlement "
+                            . "which was not provided for the IDP $name.</li>"
+                            . "<li>Please see here for more information: "
+                            . "<a href='$helpUrl' target='_blank'>$helpUrl</a>.</li>"
+                            . "<li>Logout or restart your "
+                            . "browser and attempt to login again using an IDP that "
+                            . "provides a GOCDB entitlement.</li>"
+                            . "</ul>";
+                        $HTML .= "<div style='text-align: center;'>";
+                        $HTML .= "<a href=\""
+                            . htmlspecialchars(\Factory::$properties['LOGOUTURL'])
+                            . "\"><b><font color=\"red\">Logout</font></b></a>";
+                        $HTML .= "</div>";
+                        echo ($HTML);
+                        die();
+                    }
+                }
+
+                $this->principal = $_SERVER['voPersonID'];
+                $this->userDetails = ['AuthenticationRealm' => [$provider['idp']]];
```

Comment:
This probably shouldn't just be the IdP. For EGI Check In, it should be "EGI Proxy IdP" - this should defined in the local config in the `provider` block.

Reply:
Yep Make sense. Added this in 7e7396ce3616619676f686e8d724ca6634d14964